### PR TITLE
feat: push metrics into catalog

### PR DESCRIPTION
**Pull request by tustvold**
This PR tries to address a couple of challenges associated with our metrics and catalog plumbing:

1. Gauges must be allocated once and explicitly shared between objects that wish to update the same gauge value (i.e. non-disjoint label sets)
2. Objects must explicitly decrement any contributions they made to any gauges when they are dropped
3. There is a large amount of metrics plumbing code in Db that doesn't really belong there
4. There is a large amount of metrics plumbing code in tests where we don't care about metrics
5. MUB chunk creation is unnecessarily coupled with the catalog

The first part of this PR adds a new Gauge abstraction that replaces the previous one, and is effectively a generalized version of the MemoryTracker code. This can be used with or without labels known ahead of time, the latter is more performant but often not a concern.

The second part of the PR associates a public ChunkMetrics with each Chunk that is passed in as part of the constructor. This should be default constructible, for when we don't care about metrics, and have a constructor that takes a metrics domain and any other necessary arguments when relevant.

The third part of the PR pushes all metrics out of Db into catalog where they can be tracked more accurately. Note this removes the old `catalog_chunk_size_bytes` metrics, the `catalog_chunks_mem_size_bytes` (which already existed) should be used instead, this can easily be changed if it proves controversial.

The final part of the PR special cases a chunk id of u32::MAX as not in the catalog, allowing creating MUB chunks outside the catalog and then adding them.

This PR kind of got away on me so will split some parts out of it, but I wanted to provide the final destination.

### CODE REVIEW

File: metrics/src/lib.rs [hunk 1]
```diff
@@ -51,6 +57,7 @@ pub struct MetricRegistry {
     provider: RegistryMeterProvider,
     exporter: PrometheusExporter,
     observers: Arc<ObserverCollection>,
+    gauges: Arc<Mutex<HashMap<String, Gauge>>>,
```
Comment (tustvold) [inline metrics/src/lib.rs:60]: These are registered in a single map to ensure that if two Gauges are registered with the same name, and used to publish to the same label sets, no metrics will be lost.

File: metrics/src/metrics.rs [hunk 2]
```diff
@@ -226,16 +231,25 @@ where
 /// If you want to track some notion of success, failure and latency consider
 /// using a `REDMetric` instead rather than expressing that with labels on a
 /// `Counter`.
-#[derive(Debug)]
+#[derive(Clone, Default)]
 pub struct Counter {
-    counter: OTCounter<u64>,
+    counter: Option<OTCounter<u64>>,
```
Comment (tustvold) [inline metrics/src/metrics.rs:236]: This enables default construction of `Counter`

File: metrics/src/tests.rs [hunk 3]
```diff
@@ -118,7 +118,7 @@ impl<'a> AssertionBuilder<'a> {
     }
 
     /// Returns the counter metric, allowing assertions to be applied.
-    pub fn counter(&mut self) -> Counter<'_> {
+    pub fn counter(&mut self) -> CounterAssertion<'_> {
```
Comment (tustvold) [inline metrics/src/tests.rs:121]: Renamed to avoid name clash

File: server/src/db.rs [hunk 4]
```diff
@@ -1540,9 +1236,9 @@ mod tests {
             .unwrap();
 
         // verify chunk size not increased for OS (it was in OS before unload)
-        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "os", 1921).unwrap();
+        catalog_chunk_size_bytes_metric_eq(&test_db.metric_registry, "parquet", 89).unwrap();
```
Comment (tustvold) [inline server/src/db.rs:1239]: The parquet file is currently not reporting its size correctly - https://github.com/influxdata/influxdb_iox/issues/1311

File: server/src/db.rs [hunk 5]
```diff
@@ -1761,7 +1461,7 @@ mod tests {
                 ("svr_id", "1"),
             ])
             .histogram()
-            .sample_sum_eq(3547.0)
+            .sample_sum_eq(3779.0)
```
Comment (tustvold) [inline server/src/db.rs:1464]: The reported size is a bit bigger because of the size of the metrics, previously they were separate due to the Arc, I could stick them in a Box to mitigate this, but this feels more consistent. If you're going to report metadata size, the metrics should count.

File: server/src/db/catalog/chunk.rs [hunk 6]
```diff
@@ -120,38 +107,46 @@ macro_rules! unexpected_state {
...
-    /// 3. a new open chunk is created using the buffer
-    /// 4. a write is recorded (see [`record_write`](Self::record_write))
     pub(crate) fn new_open(
-        batch: TableBatch<'_>,
         partition_key: impl AsRef<str>,
-        id: u32,
-        clock_value: ClockValue,
-        server_id: ServerId,
-        memory_registry: &MemRegistry,
+        chunk: mutable_buffer::chunk::Chunk,
```
Comment (tustvold) [inline server/src/db/catalog/chunk.rs:125]: The catalog no longer constructs the MUB chunk. This makes it more consistent with how other chunks are handled, and more specifically avoided a confusing situation where you pass in a metrics domain for the mutable chunk, to the catalog, which just felt odd

File: server/src/db/catalog/chunk.rs [hunk 7]
```diff
@@ -366,13 +364,27 @@ impl Chunk {
     /// Set the chunk to the Moving state, returning a handle to the underlying
     /// storage
     pub fn set_moving(&mut self) -> Result<Arc<MBChunk>> {
+        // This ensures the closing logic is consistent but doesn't break code that
```
Comment (tustvold) [inline server/src/db/catalog/chunk.rs:367]: Previously `time_closed` would never actually get set

File: tracker/src/mem.rs [hunk 8]
```diff
@@ -1,117 +0,0 @@
-//! This module contains a basic memory tracking system
```
Comment (tustvold) [inline tracker/src/mem.rs:1]: This construction has been replaced by GaugeValue

File: server/src/db.rs [hunk 9]
```diff
@@ -1204,41 +917,35 @@ impl Db {
...
-                            );
-
-                            check_chunk_closed(&mut *chunk, mutable_size_threshold, &self.metrics);
+                            check_chunk_closed(&mut *chunk, mutable_size_threshold);
                         }
                         None => {
-                            let new_chunk = partition
-                                .create_open_chunk(
-                                    table_batch,
+                            let metrics = self.metrics_registry.register_domain("mutable_buffer");
```
Comment (tustvold) [inline server/src/db.rs:923]: This shows how MUB chunks are constructed outside the catalog and then inserted into it when ready

File: read_buffer/src/chunk.rs [hunk 10]
```diff
@@ -696,45 +627,42 @@ impl ChunkMetrics {
...
-            column_bytes_total: domain.register_gauge_metric_with_labels(
-                "column",
-                Some("bytes"),
-                "The number of bytes used by all columns in the Read Buffer",
-                vec![metrics::KeyValue::new("db", db)],
-            ),
-        }
-    }
-}
+    // Updates column storage statistics for the Read Buffer.
```
Comment (tustvold) [inline read_buffer/src/chunk.rs:630]: This code is just moved

File: server/src/db.rs [hunk 11]
```diff
@@ -421,52 +321,22 @@ impl Db {
...
-
-        let memory_registries = Default::default();
-        domain.register_observer(None, &[], &memory_registries);
+        let metric_labels = vec![
+            KeyValue::new("db_name", db_name.to_string()),
+            KeyValue::new("svr_id", format!("{}", server_id)),
+        ];
 
-        let db_metrics = DbMetrics {
-            catalog_chunks: domain.register_counter_metric_with_labels(
```
Comment (tustvold) [inline server/src/db.rs:439]: This metric has been pushed into the catalog

File: server/src/db.rs [hunk 12]
```diff
@@ -421,52 +321,22 @@ impl Db {
...
-                "chunks",
-                None,
-                "In-memory chunks created in various life-cycle stages",
-                vec![],
-            ),
-            catalog_immutable_chunk_bytes: domain
```
Comment (tustvold) [inline server/src/db.rs:445]: This metric has also been pushed into the catalog

File: server/src/db.rs [hunk 13]
```diff
@@ -421,52 +321,22 @@ impl Db {
...
-                .register_histogram_metric(
-                    "chunk_creation",
-                    "size",
-                    "bytes",
-                    "The new size of an immutable chunk",
-                )
-                .init(),
-            catalog_chunk_bytes: domain.register_gauge_metric_with_labels(
```
Comment (tustvold) [inline server/src/db.rs:453]: This metric has been removed

File: server/src/db.rs [hunk 14]
```diff
@@ -421,52 +321,22 @@ impl Db {
...
-                "chunk_size",
-                Some("bytes"),
-                "The size in bytes of all chunks",
-                vec![],
-            ),
-            read_buffer_chunk_metrics: Arc::new(read_buffer::ChunkMetrics::new_with_db(
```
Comment (tustvold) [inline server/src/db.rs:459]: This workaround is no longer necessary due to the new gauages

**Comment by tustvold** [comment]
So I initially thought I'd split this up, but the changes are so intertwined that it is actually quite complex to do so. I can't push the metrics into the catalog until the Gauge's are fixed, and I can't change the behavior of the Gauage's without changing all the call-sites.

I dunno, let me know if you would prefer I did split this up

File: metrics/src/gauge.rs [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+/// A `Gauge` allows tracking multiple usize values by label set
+///
+/// Metrics can be recorded directly on the Gauge or when the labels are
+/// known ahead of time a `GaugeValue` can be obtained with `Gauge::gauge_value`
+///
+/// When a `Gauge` is dropped any contributions it made to any label sets
+/// will be deducted
+#[derive(Debug, Default)]
+pub struct Gauge {
+    shared: Arc<GaugeShared>,
```
Comment (NGA-TRAN) [inline metrics/src/gauge.rs:20]: Maybe a comment explaining what `shared` is used for.

File: metrics/src/gauge.rs [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+        assert_eq!(t1.get_total(), 400);
+        assert_eq!(t2.get_total(), 400);
+        assert_eq!(start.get_total(), 400);
+
+        t2.set(400);
+        assert_eq!(t1.get_total(), 500);
+        assert_eq!(t2.get_total(), 500);
+        assert_eq!(start.get_total(), 500);
+
+        std::mem::drop(t2);
```
Comment (alamb) [inline metrics/src/gauge.rs:223]: 👍  good test -- it illustrates the user of shared gauges well.

File: metrics/src/lib.rs (hunk 1 above)
Comment (alamb) [reply metrics/src/lib.rs:60]: a list of gauges makes sense to me, FWIW

File: metrics/src/lib.rs [hunk 17]
```diff
@@ -341,40 +364,33 @@ impl Domain {
...
+            }
+        };
+        gauge.set_labels([&self.default_labels[..], &default_labels].concat());
+        gauge
     }
 
+    /// An observer can be used to provide asynchronous fetching of values from an object
+    ///
+    /// NB: If you register multiple observers with the same measurement name, they MUST
+    /// publish to disjoint tag sets
```
Comment (alamb) [inline metrics/src/lib.rs:393]: might be worth clarifying what happens otherwise ("otherwise only an arbitrary observer will be interrogated asynchronously"??)

File: metrics/src/gauge.rs [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+                f(vacant.insert(observer))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+struct GaugeShared {
+    /// Maps a set of tags to a gauge and its tags
+    values: DashMap<String, (GaugeValue, Vec<KeyValue>)>,
```
Comment (NGA-TRAN) [inline metrics/src/gauge.rs:97]: Because I am new to metrics it is very hard for me to understand this data structure. Adding comments here  explaining what it is with an example will make it clear for reviewers and coders who need to touch it later

File: metrics/src/gauge.rs [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+            .value()
+            .0
+            .clone()
+    }
+}
+
+#[derive(Debug, Default)]
+struct GaugeValueShared {
+    /// The total bytes across all associated `GaugeValueHandle`s
+    bytes: AtomicUsize,
```
Comment (alamb) [inline metrics/src/gauge.rs:128]: is is really the total number of 'bytes'? Or is this the 'shared value backing this gauge'?

File: metrics/src/gauge.rs [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+        assert_eq!(t1.get_total(), 500);
+        assert_eq!(t2.get_total(), 500);
+        assert_eq!(start.get_total(), 500);
+
+        std::mem::drop(t2);
+        assert_eq!(t1.get_total(), 100);
+        assert_eq!(start.get_total(), 100);
+
+        std::mem::drop(t1);
+        assert_eq!(start.get_total(), 0);
```
Comment (NGA-TRAN) [inline metrics/src/gauge.rs:228]: This test is very helpful. So all the clones share the same pool but they can only set value of themselves which is a part of the pool?

File: metrics/src/gauge.rs [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    #[test]
+    fn test_mixed() {
+        let mut gauge = Gauge::new(vec![KeyValue::new("foo", "bar")]);
```
Comment (NGA-TRAN) [inline metrics/src/gauge.rs:233]: So we have a use case that the gauge is a key-value pair?

File: server/src/db/catalog.rs [hunk 22]
```diff
@@ -148,23 +141,9 @@ impl Catalog {
...
-                    &[
-                        KeyValue::new("lock", "chunk"),
-                        KeyValue::new("partition", entry.key().clone()),
-                    ],
-                    &lock_tracker,
-                );
-
-                let partition = Partition::new(entry.key(), lock_tracker);
-                let partition = Arc::new(self.lock_tracker.new_lock(partition));
+                let partition_metrics = self.metrics.new_partition_metrics(entry.key().to_string());
```
Comment (alamb) [inline server/src/db/catalog.rs:144]: well, that is a lot nicer 👍

File: metrics/src/lib.rs (hunk 1 above)
Comment (NGA-TRAN) [reply metrics/src/lib.rs:60]: > Gauges are registered with the same name, and used to publish to the same label sets, no metrics will be lost

Dumb question: How do we distinguish them?

File: metrics/src/lib.rs [hunk 23]
```diff
@@ -1277,4 +1294,118 @@ mod test {
...
+            .unwrap();
+
+        std::mem::drop(gauge1);
+
+        reg.has_metric_family("catalog_read_buffer")
+            .with_labels(&[("foo", "bar"), ("enc", "foo")])
+            .gauge()
+            .eq(0.)
+            .unwrap();
+    }
```
Comment (NGA-TRAN) [inline metrics/src/lib.rs:1410]: Should we add tests for `dec`? What happen if `dec` leads to negative? error will be thrown?

File: read_buffer/src/chunk.rs [hunk 24]
```diff
@@ -1098,9 +1023,7 @@ mod test {
 
     #[test]
     fn table_summaries() {
-        let reg = metrics::TestMetricRegistry::new(Arc::new(metrics::MetricRegistry::new()));
```
Comment (alamb) [inline read_buffer/src/chunk.rs:1101]: these tests sure are a lot nicer 👍

**Review by alamb — APPROVED** [review]
Overall I think this code makes a lot of sense and I found it a joy to read (if a bit lengthy). 👍 

I suspect this may conflict with https://github.com/influxdata/influxdb_iox/pull/1482 (perhaps badly) fyi @crepererum 

The only thing I think we should verify before merging is if using `usize` for gauges (rather than `f64` as the old implementation did) is ok.  I realize the approach in this PR requires using atomics which aren't supported for floating point numbers.

I am not enough of an expert user to know if using floating point values for gauges is important or not, but it is probably worth asking @e-dard  or @jacobmarble .

File: server/src/db.rs [hunk 25]
```diff
@@ -1865,7 +1601,7 @@ mod tests {
                 ("svr_id", "10"),
             ])
             .histogram()
-            .sample_sum_eq(1921.0)
+            .sample_sum_eq(2153.0)
```
Comment (NGA-TRAN) [inline server/src/db.rs:1604]: More bytes registered and hence accurate, now?

File: server/src/db/catalog/metrics.rs [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+        self.lock_tracker.new_lock(t)
+    }
+
+    pub(super) fn new_partition_metrics(&self, partition_key: impl ToString) -> PartitionMetrics {
+        // Lock tracker for chunk-level locks
+        let lock_tracker = Default::default();
+        self.metrics_domain.register_observer(
+            None,
+            &[
+                KeyValue::new("lock", "chunk"),
```
Comment (NGA-TRAN) [inline server/src/db/catalog/metrics.rs:48]: Ah, so the use case of the key-value is here

**Review by NGA-TRAN — APPROVED** [review]
All the changes relate to the restructuring of metrics so they make sense to me. Nice work, @tustvold

**Review by carols10cents — COMMENTED** [review]
I have some thoughts and questions, overall I like where it's moving the responsibilities of metrics tracking though!

File: metrics/src/gauge.rs (hunk 19 above)
Comment (tustvold) [reply metrics/src/gauge.rs:128]: Heh - that's a holdover from when this used to be the MemoryTracker :D

File: metrics/src/gauge.rs (hunk 21 above)
Comment (tustvold) [reply metrics/src/gauge.rs:233]: This is just setting a default labels that will be included with any observations made by through this Gauge instance

File: server/src/db/catalog/metrics.rs (hunk 26 above)
Comment (tustvold) [reply server/src/db/catalog/metrics.rs:48]: This is registering a lock tracker, and configuring it such that any observations it makes will have the labels `("lock", "chunk")` and `"partition", partition_key) associated with them, in addition to whatever they might already have

File: mutable_buffer/src/chunk.rs [hunk 27]
```diff
@@ -45,6 +48,18 @@ pub enum Error {
...
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
+#[derive(Debug, Default)]
+pub struct ChunkMetrics {
+    /// keep track of memory used by chunk
+    memory_bytes: GaugeValue,
+}
+
+impl ChunkMetrics {
+    pub fn new(_metrics: &metrics::Domain, memory_bytes: GaugeValue) -> Self {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply mutable_buffer/src/chunk.rs:58]: If/when the mutable buffer has metrics of its own to report it will use it to register its own counters, gauges, etc...

File: mutable_buffer/src/chunk.rs [hunk 28]
```diff
@@ -161,6 +176,19 @@ impl Chunk {
         self.id
     }
 
+    pub fn set_id(&mut self, id: u32) {
+        assert_eq!(
+            self.id, INVALID_CHUNK_ID,
+            "cannot change ID once it has been set"
+        );
+        self.id = id;
```
[the remark this answers is not in the document]
Comment (tustvold) [reply mutable_buffer/src/chunk.rs:184]: I was being lazy and trying to avoid changing the signature of new :sweat_smile: I'll just do it properly as you suggest

File: metrics/src/gauge.rs [hunk 29]
```diff
@@ -0,0 +1,259 @@
...
+    values: DashMap<String, (GaugeValue, Vec<KeyValue>)>,
+}
+
+impl GaugeShared {
+    fn encode_labels(&self, labels: impl IntoIterator<Item = KeyValue>) -> (String, LabelSet) {
+        let set = LabelSet::from_labels(labels);
+        let encoded = set.encoded(Some(&DefaultLabelEncoder));
+
+        (encoded, set)
+    }
```
[the remark this answers is not in the document]
Comment (tustvold) [reply metrics/src/gauge.rs:106]: Legacy holdover, will fix

File: metrics/src/gauge.rs [hunk 30]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+/// A `GaugeValue` is a single measurement associated with a `Gauge`
+///
+/// When the `GaugeValue` is dropped any contribution it made to the total
+/// will be deducted from the `GaugeValue` total
+#[derive(Debug, Default)]
+pub struct GaugeValue {
+    shared: Arc<GaugeValueShared>,
+    /// The locally observed value
```
Comment (e-dard) [inline metrics/src/gauge.rs:138]: If this field isn't public does it make sense for the doc comment to be using `///`?

File: metrics/src/gauge.rs [hunk 31]
```diff
@@ -0,0 +1,259 @@
...
+    local: usize,
+}
+
+impl Clone for GaugeValue {
+    fn clone(&self) -> Self {
```
Comment (e-dard) [inline metrics/src/gauge.rs:143]: I don't really consider this a `clone`. It looks to me like you're constructing a new `GaugeValue` using some existing shared state to me.

I get that you're implenting `Clone` so duh it needs to be called `clone`  but it doesn't quite seem like the right name for what's going on here :-)

If a `GaugeValue` is useless without a pointer to a `GaugeValueShared` the it sort of feels like this should just be a `new(shared: Arc<GaugeValueShared>)` constructor 🤷

File: metrics/src/metrics.rs [hunk 32]
```diff
@@ -248,6 +264,11 @@ impl Counter {
     /// Increase the count by `value` and associate the observation with the
     /// provided labels.
     pub fn add_with_labels(&self, value: u64, labels: &[KeyValue]) {
+        let counter = match self.counter.as_ref() {
+            Some(counter) => counter,
+            None => return,
```
[the remark this answers is not in the document]
Comment (e-dard) [reply metrics/src/metrics.rs:269]: Yeah I agree. I don't think metrics should be no-ops when things are called on them. If you don't care metrics in tests then don't assert anything against them. I think the metrics should always be being correctly exercised.

**Review by e-dard — APPROVED** [review]
@tustvold I'm loving how the metrics crate is being iterated on and vastly improved :-) This is look great. I had a couple of comments about semantics but I think the code is good.

File: metrics/src/metrics.rs (hunk 32 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply metrics/src/metrics.rs:269]: I dunno, perhaps it is because I wrote it, but that's exactly what I'd expect Default to do. If you aren't passing in a registry to record to, I at least would expect metrics recording to be a no-op. I'd really prefer not to use cfg_attr, as not only would that not work for benchmarks, but also pollute all the code with cfg_attr.

I debated adding a new_noop() method, but then all the downstreams need to implement their own version of it, and it's just a load of extra code. I dunno I'm not sure how others feel about this...

File: metrics/src/gauge.rs (hunk 30 above)
Comment (tustvold) [reply metrics/src/gauge.rs:138]: I thought it was best practice to always use `///` on fields, structs, etc... in case rustdoc is run with show private, and just for general consistency? Is there a reason not to?

**Comment by tustvold** [comment]
Thank you all for the reviews, I think I've incorporated your comments. 

The only one I haven't is the question around Default and no-op, I'd like to discuss this at standup. I have a concern that the metrics setup requires passing metrics instruments from potentially multiple domains into the `ChunkMetrics` constructor, in order to have them report as the same metric name. An example of this would be the catalog metrics, but this would apply to any database-level measurement. This will quickly get unwieldy if every test has to properly construct all of them, and to be completely honest, I don't understand the benefit. We should have integration checks of the metrics plumbing, as with any system, I'm not sure why the unit tests for a component should be testing anything to do with metrics plumbing...

File: metrics/src/gauge.rs [hunk 33]
```diff
@@ -0,0 +1,265 @@
...
+    ///
+    /// The key is a sorted encoding of this label set, that reconciles
+    /// duplicates in the same manner as OpenTelemetry (it uses the same encoder)
+    /// which appears to be last-writer-wins
+    values: DashMap<String, (GaugeValue, Vec<KeyValue>)>,
+}
+
+impl GaugeState {
+    /// Visits the totals for all recorded label sets
+    pub fn visit_values(&self, f: impl Fn(usize, &Vec<KeyValue>)) {
```
Comment (shepmaster) [inline metrics/src/gauge.rs:89]: `&Vec<T>` is usually better written as `&[T]`

File: metrics/src/gauge.rs (hunk 30 above)
Comment (e-dard) [reply metrics/src/gauge.rs:158]: Very cool. Glad I asked the question now as I learnt something :-)

**Review by alamb — APPROVED** [review]
I reviewed https://github.com/influxdata/influxdb_iox/pull/1488/commits/1dead86c1a08559c59a3dcf7b65086584e12e1d9

And the change to use `new_unregistered` instead of `Default::default()` makes a lot of sense to me 👍
